### PR TITLE
Create Address spec

### PR DESCRIPTION
**Pull request by frrist**
### Description
Initial write up of Address definition

### CODE REVIEW

File: address.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    return digest[len(digest)-CKSM_LEN:]
+}
+
+// validChecksum returns true if cksm is valid for data and false otherwise.
+func validChecksum(data, cksm []byte) bool {
+    digest := Checksum(data)
+    return digest == cksm
+}
+```
+
```
[the remark this answers is not in the document]
Comment (frrist) [reply address.md]: Sounds good, I have added an interface to match the implementation.

**Comment by frrist** [comment]
Couple questions I'd like to talk about:
- Are we alright with base58 encoding or should we use something that is case insensitive instead, such as base32?
- Do we want addresses to use multibase so that we may represent the same address with many different encodings?

**Comment by bvohaska** [comment]
Ping @whyrusleeping on @frrist comment about baseX encodings. I have no strong feelings either way. What is the industry standard on this? Since this effort follows Multiformats, is there a prefered base or should we go with multibase?

**Comment by dignifiedquire** [comment]
we should use base32, to avoid having sth that is case sensitive. the issues are well documented about this. multibase adds overhead so that depends on if we are okay with that. also ipfs is switching to base32 as the default

File: address.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+An exampple implementation of this interface is shown below:
+
+```go
+type Address struct {
+    // 0: main-net
+    // 1: test-net
+    network byte
```
Comment (dignifiedquire) [inline address.md]: when did we decide to stop differentiating between testnets?

File: address.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    
+    return Address{
+        network: ntwk,
+        typ: raw[0],
+        data: raw[1: len(raw)- CKSM_LEN]
+    }
+}
+
+// Checksum is the last CKSM_LEN bytes the sha256 of Address Type and its data.
+func Checksum(a Address) [CKSM_LEN]byte {
```
Comment (dignifiedquire) [inline address.md]: why are we using sha256 suddenly as checksums? we had a proper scheme with error detections in the previous spec version

File: address.md (hunk 2 above)
Comment (frrist) [reply address.md]: I don't think we have decided that, what do you mean?

File: address.md (hunk 3 above)
Comment (frrist) [reply address.md]: What scheme were we previously using? I assume you mean [this one](https://github.com/filecoin-project/go-filecoin/blob/efdc601098fec6d71f1543327090c1e468d0836b/address/address.go#L327-L373), if so we continue to use it -- `sha256` was chosen here for simplicity in writing as I wasn't sure on the details of the aforementioned code.

File: address.md (hunk 2 above)
Comment (whyrusleeping) [reply address.md]: @dignifiedquire I don't really think its useful. If many people disagree with that, we could put it in the user visible encoding, but not include it in the bytes representation.

File: address.md (hunk 3 above)
Comment (whyrusleeping) [reply address.md]: @dignifiedquire the previous scheme was just a complicated checksum, its far simpler to use a few bytes from a well known hash function.

File: address.md [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+Learning from the Ethereum:
+
+* Hex encoding is inefficient with respect to storage. Representing addresses as Hex increases the address size by 100%.
+* Checksums require mixed cases, resulting in the same issues as above.
+
+## Specification
+
+A Filecoin address is a byte string consisting of:
+* The **letter 'f'** to identify the address as belonging to a Filecoin network.
```
Comment (frrist) [inline address.md]: It is unclear to me if this `f` exists in addition to the network indicator or if it is a network indicator. The current implementation treats f as a network indicator.

File: address.md [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+	SECP256K1 = 0
+	ID = 1
+	Actor = 2
+	BLS = 3
+)
+```
+
+###### Protocol 0: libsecpk1 Elliptic Curve Public Keys
+
+**Protocol 0** addresses represent public encryption keys. These addresses represent Filecoin user accounts and are likely to be transmitted outside the network over noisy channels (e.g. hand written on a napkin). As such, they require robust error detection. The multi-base encoded key contains a hash and a checksum. The hash is a [Blake2b 160](https://blake2.net/) hash of the public key. The checksum is a six character BCH checksum. The algorithm details can be found in Appendix - 9.2. Algorithms - Checksums.
```
Comment (frrist) [inline address.md]: I believe we either need to add a second protocol for the case of Actors without public keys, or we need to state that the address represent public encryption keys **or** "random data" used to generate an address. I think "random data" is a pretty poor description, thoughts?

Comment (frrist) [reply address.md]: Also, I believe we decided that we are **not** using multi-base encoding, correct?

Comment (frrist) [inline address.md]: I believe we decided that we are **not** using multi-base encoding, correct?

Comment (frrist) [inline address.md]: Should we have a bit here on secp256k1 public key recovery since we are using the hash of the public key?

**Comment by whyrusleeping** [comment]
Just to state it clearly, we should use a truncated common hash (sha256 or blake2) as the checksum, its much simpler for others to implement, and easier to reason about, and also doesnt have to be targeted to a particular size of input.

**Comment by bvohaska** [comment]
@whyrusleeping: BCH encoding will allow a receiver to (1) check if the data is correct (2) fix some set of errors if there was a problem. Truncated hashes will provide some data integrity but no correction. BCH codes are also blazing fast (not that we expect this to be an issue).

As far as reasonability, BCH should be just as easy to reason about as hashing for error detection. We can also provide some clearer context as it’s spec’d out a bit more. I believe BCH codes are also used in QR Codes so there is also some external reference for devs. Putting addresses into QR codes could be a pretty cool feature in any case.

**Comment by whyrusleeping** [comment]
I posit that the error correction is not super useful, people are not going to be typing these, they will be copy pasted. If someone makes an error in a copy pasted address, and we're just okay with it because the error gets corrected, then that address is assumed to be correct, when its not. This can lead to multiple unique strings that actually correct to the same address.

As far as "easy" goes, the fact that we have to implement BCH codes ourselves and its not just `import "crypto/sha"` makes it far far less easy. The speed of hashing is really not a concern, as its only done when presenting things to the user.

**Comment by dignifiedquire** [comment]
implementation wise we already have it implemented and used in the current address format, so there is not really much to do there.

File: address.md (hunk 3 above)
Comment (frrist) [reply address.md]: Based on other discussion we have had we are going with the first 4 bytes of the blake2b hash of an address

File: address.md (hunk 2 above)
Comment (frrist) [reply address.md]: We will have support for Mainnet and Testnet in the initial address spec

File: address.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+- `1` : Testnet
+
+An example description in golang:
+
+```go
+// String Prefix
+Mainnet_Prefix = 'f'
+Testnet_Prefix = 't'
+
+type Protocol = byte
```
Comment (rkowalick) [inline address.md]: this should be `Network`

File: address.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    ID Protocol = iota
+    SECP256K1
+    Actor
+    BLS
+)
+```
+
+###### Protocol 0: IDs
+
+**Protocol 0** addresses are simple ids. These addresses are expected to remain within the network. All actors have a numeric ID even if they don't have public keys. The key part of an ID address is the decimal representation of the id only characters '0'-'9'. IDs are not hashed and do not have a checksum.
```
Comment (rkowalick) [inline address.md]: This sentence is confusing to me:
```
The key part of an ID address is the decimal representation of the id only characters '0'-'9'.
```

File: address.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+**Protocol 1** addresses represent secp256k1 public encryption keys. The payload field contains the [Blake2b 160](https://blake2.net/) hash of the public key. 
+
+###### Protocol 2: Actor
+
+**Protocol 2** addresses representing an Actor. The payload field contains the [Blake2b 160](https://blake2.net/) hash of meaningful data produced as a result of creating the actor. 
+
+###### Protocol 3: BLS
+
+**Protocol 3** addresses represent BLS public encryptions keys. The payload filed contains the BLS public key.
```
Comment (rkowalick) [inline address.md]: typo `filed` => `field`

File: address.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+#### Payload
+
+The payload represents the data specified by the protocol. All payloads excecpt the payload of the ID protocol are base32 encoded when seralized to their human readable format.
```
Comment (rkowalick) [inline address.md]: typo `excecpt` => `except`

File: address.md [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+}
+```
+##### New()
+
+New returns an Address for the specified network and protocol encapsulating corresponding payload. New fails for unknown network or protocol.
+
+```go
+func New(network byte, protocol byte, payload []byte) Address {
+	if network != Mainnet && network != Testnet {
+        Fatal(ErrUnknownNetwork)
```
Comment (rkowalick) [inline address.md]: indent

File: address.md [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+	}
+	if protocol < SECP256K1 || protocol > BLS {
+        Fatal(ErrUnknownType)
```
Comment (rkowalick) [inline address.md]: indent

File: address.md [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+	}
+    return Address{
+        Network:  network,
```
Comment (rkowalick) [inline address.md]: `Network` and `Payload` are capitalized here and lowercase in the struct definition.

File: address.md [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
+
+Encodes an Address as a string, converting the network byte to the corresponding prefix and encoding the payload to [base32](https://tools.ietf.org/html/rfc4648).
+
+```go
+func Encode(a Address) string {
```
Comment (rkowalick) [inline address.md]: indentation in this func is kind of jank

File: address.md [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+    switch a[:2] {
+    case Mainnet_Prefix:
+        ntwk = Mainnet
+    case Testnet_Prefix:
+    	ntwk = Testnet
+    default:
+        Fatal(ErrUnknownNetwork)
+    }
+    
+    protocol := a[3]
```
Comment (rkowalick) [inline address.md]: I think this should be `a[2]`

File: address.md [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+    
+    if protocol == ID {
+        return Address{
+            network: ntwk,
+            protocol: protocol
+            payload: payload
+        }
+    }
+    
+    cksm := payload[len(payload)-CksmLen:len(payload)]
```
Comment (rkowalick) [inline address.md]: can be simplified to `cksm := payload[len(payload)-CksmLen:]

Comment (rkowalick) [reply address.md]: Also `ChksmLen` is used before it is defined.

File: address.md [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+```
+
+##### Checksum()
+
+Checksum produces a byte array of the blake2b-16 checksum of an address, returning the first 4 bytes of the digest.
+
+```go
+const CksmLen = 4
+
+func Checksum(a Address) [CKSM_LEN]byte {
```
Comment (rkowalick) [inline address.md]: The const above is not clearly being used here. Also, are we hashing the entire address struct in this function?

File: address.md [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+| 1 byte  |  1 byte  | n bytes |  4 bytes |
+```
+
+An example of a Filecoin address in golang:
+
+```go
+type Address struct {
+    // 0: main-net
+    // 1: test-net
+    network byte
```
Comment (dignifiedquire) [inline address.md]: unclear if we need to actually store this here

File: address.md [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+    payload []byte
+}
+```
+
+#### Network Indicator
+
+The **network** byte identifies the network the address belongs to:
+
+- `0` : Mainnet
+- `1` : Testnet
```
Comment (dignifiedquire) [inline address.md]: We don't need byte identifiers for these, these are just relevant to encoded addresses

File: address.md (hunk 7 above)
Comment (dignifiedquire) [inline address.md]: maybe saying that they are encoded as decimal numbers is more straightforward

File: address.md [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+
+**Protocol 3** addresses represent BLS public encryptions keys. The payload filed contains the BLS public key.
+
+#### Payload
+
+The payload represents the data specified by the protocol. All payloads excecpt the payload of the ID protocol are base32 encoded when seralized to their human readable format.
+
+#### Checksum
+
+The checksum is calculated using the first 4 bytes of the blake2b-160 hash of the addresses network, protocol, and payload bytes. Checksums are only used when seralizing and deseralizing an address to its human readable format.
```
Comment (dignifiedquire) [inline address.md]: the network should not be part of the hash if we hash the byte format

File: address.md [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+
+```go
+type Address interface {
+    New(n byte, p byte, payload []byte) Address
+    Encode(a Adress) string
+    Decode(s string) Address
+    Checksum(a Address) []byte
+    Validate(a Address) bool
+}
+```
```
Comment (dignifiedquire) [inline address.md]: no error handling?

File: address.md [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+
+##### Encode()
+
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
+
+Encodes an Address as a string, converting the network byte to the corresponding prefix and encoding the payload to [base32](https://tools.ietf.org/html/rfc4648).
```
Comment (dignifiedquire) [inline address.md]: missing the lower case part

File: address.md [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+Decode an Address from a string, converting the network prefix and protocol prefix to their corresponding byte values, decoding the payload from base32, and validating the checksum.
+
+```go
+func DecodeString(a string) Address {
+    if len(a) < 4 {
+        Fatal(ErrInvalidLength)
+    }
+    
+    var ntwk byte
+    switch a[:2] {
```
Comment (dignifiedquire) [inline address.md]: this should be `a[0]`

File: address.md (hunk 14 above)
Comment (dignifiedquire) [reply address.md]: actually this should be `a[1]`

File: address.md [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+    case Mainnet_Prefix:
+        ntwk = Mainnet
+    case Testnet_Prefix:
+    	ntwk = Testnet
+    default:
+        Fatal(ErrUnknownNetwork)
+    }
+    
+    protocol := a[3]
+    payload := a[3:]
```
Comment (dignifiedquire) [inline address.md]: this should be `a[2:]`

File: address.md [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+
+##### Checksum()
+
+Checksum produces a byte array of the blake2b-16 checksum of an address, returning the first 4 bytes of the digest.
+
+```go
+const CksmLen = 4
+
+func Checksum(a Address) [CKSM_LEN]byte {
+    digest = blake2b(a)
```
Comment (dignifiedquire) [inline address.md]: did you see my comment regarding using `Blake2b-4` instead?

File: data-structures.md [hunk 25]
```diff
@@ -4,6 +4,27 @@ This document serves as an entry point for understanding all of the data structu
...
+type Address struct {
+    // 0: main-net
+    // 1: test-net
+    network leb128
+
+    // 0: SECP256K1 Public Key
+    // 1: Actor
+    // 2: BLS Public Key
+    // 3: ID
+    typ leb128
```
Comment (dignifiedquire) [inline data-structures.md]: these are bytes, not leb128, according to our last agreement

File: data-structures.md [hunk 26]
```diff
@@ -4,6 +4,27 @@ This document serves as an entry point for understanding all of the data structu
...
 
+## Address
+
+An address is an identifier that refers to an actor in the Filecoin state. All actors (miner actors, the storage market actor, account actors) have an address. An address encodes information about the network it belongs to, the type of data it contains, the data itself, and depending on the type, a checksum.
+
+```go
+type Address struct {
+    // 0: main-net
+    // 1: test-net
+    network leb128
```
Comment (dignifiedquire) [inline data-structures.md]: these are bytes, not leb128, according to our last agreement

File: address.md (hunk 17 above)
Comment (whyrusleeping) [reply address.md]: It doesnt hurt, might make some things easier (that way addresses don’t have to depend on some other thing to get this info)

File: address.md [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+    protocol byte
+    
+    // raw bytes containing the data associated with protocol
+    payload []byte
+}
+```
+
+#### Network Indicator
+
+The **network** byte identifies the network the address belongs to:
```
Comment (whyrusleeping) [inline address.md]: i’d call it a network prefix instead of a network btye

File: address.md (hunk 7 above)
Comment (whyrusleeping) [reply address.md]: Yeah, worth just saying its encoded base 10

File: address.md [hunk 28]
```diff
@@ -0,0 +1,262 @@
...
+)
+```
+
+###### Protocol 0: IDs
+
+**Protocol 0** addresses are simple ids. These addresses are expected to remain within the network. All actors have a numeric ID even if they don't have public keys. The key part of an ID address is the decimal representation of the id only characters '0'-'9'. IDs are not hashed and do not have a checksum.
+
+###### Protocol 1: libsecpk1 Elliptic Curve Public Keys
+
+**Protocol 1** addresses represent secp256k1 public encryption keys. The payload field contains the [Blake2b 160](https://blake2.net/) hash of the public key.
```
Comment (whyrusleeping) [inline address.md]: mention how these are encoded for human readability

Comment (whyrusleeping) [reply address.md]: oh i see, its mentioned later

File: address.md [hunk 29]
```diff
@@ -0,0 +1,303 @@
...
+
+###### Protocol 2: Actor
+
+**Protocol 2** addresses representing an Actor. The payload field contains the [Blake2b 160](https://blake2.net/) hash of meaningful data produced as a result of creating the actor. 
+
+```
+|------------|----------|---------------------|----------|
+|  network   | protocol |      payload        | checksum |
+|------------|----------|---------------------|----------|
+| 'f' or 't' |     2    |  blake2b-160(TODO)  |  4 bytes |
```
Comment (frrist) [inline address.md]: Need to address this TODO cc @whyrusleeping, what values are we hasing here?

File: address.md [hunk 30]
```diff
@@ -0,0 +1,303 @@
...
+    ID Protocol = iota
+    SECP256K1
+    Actor
+    BLS
+)
+```
+
+###### Protocol 0: IDs
+
+**Protocol 0** addresses are simple IDs. These addresses are expected to remain within the network. All actors have a numeric ID even if they don't have public keys. The payload of an ID address is encoded as decimal numbers '0'-'9'. IDs are not hashed and do not have a checksum.
```
Comment (frrist) [inline address.md]: What encoding are we using for ID's?
I believe the following are all valid filecoin addresses under this current definition, correct?
`f00`
`f010`
`f0123`

File: address.md (hunk 20 above)
Comment (frrist) [reply address.md]: We use `Fatal` when there is an error, basically a panic. Validation has been added here: https://github.com/filecoin-project/specs/pull/135/files#diff-c6ea617359921b8070df6dd947c8af6cR257

File: address.md [hunk 31]
```diff
@@ -0,0 +1,303 @@
...
+                  base32[................................]
+```
+
+#### Payload
+
+The payload represents the data specified by the protocol. All payloads except the payload of the ID protocol are [base32](https://tools.ietf.org/html/rfc4648) encoded when seralized to their human readable format. The base32 encoding uses the lowercase alphabet.
+
+#### Checksum
+
+The checksum is calculated using blake2b-4 hash of the addresses protocol, and payload bytes. Checksums are base32 encoded  and only used when encoding and decoding an address to its human readable format -- a string. Checksums will never appear on chain.
```
Comment (frrist) [inline address.md]: @dignifiedquire we have explicitly mentioned blake2b-4 here.

File: address.md [hunk 32]
```diff
@@ -0,0 +1,303 @@
...
+##### Checksum()
+
+Checksum produces a byte array of the blake2b-4 checksum of an addresses protocol and payload.
+
+```go
+const CksmLen = 4
+
+func Checksum(a Address) [CksmLen]byte {
+    digest = blake2b(a.Protocol + a.Payload)
+    return digest[:4]
```
Comment (dignifiedquire) [inline address.md]: @frrist this is not blake2b-4, the `4` needs to be passed to the hashing function as an argument, like here: https://github.com/filecoin-project/go-filecoin/blob/6914061e22bb8725182e5496b1286c914b5b9c26/address/address.go#L15

Comment (frrist) [reply address.md]: Ahh yes, I missed this bit. Seeing as this code isn't expected to compile and is serving here as an example, do we need to add the hash function configuration code? or could we just do something like
`return blake2b4(a.Protocol + a.Payload)`

Comment (dignifiedquire) [reply address.md]: you can do that, and add a comment maybe saying what this is supposed to mean?

Comment (frrist) [reply address.md]: Sounds good to me, will do that.

File: address.md [hunk 33]
```diff
@@ -0,0 +1,300 @@
...
+| protocol | payload |
+|----------|---------|
+|  1 byte  | n bytes |
+```
+
+A Filecoin address encoded to a string consistes of:
+* A **network prefix** character that identifies the network the address belongs to.
+* A **protocol indicator** byte that identifies the type and version of this address.
+* A **payload** used to uniquely identify the actor according to the protocol..
+* A **checksum** used to validet the address.
```
Comment (whyrusleeping) [inline address.md]: typo on validate, and also maybe note that the checksum isnt always present?

File: address.md (hunk 30 above)
Comment (whyrusleeping) [reply address.md]: What do you mean by 'these addresses are expected to remain within the network.' ?

File: address.md [hunk 34]
```diff
@@ -0,0 +1,300 @@
...
+|    0     | '0'-'9' |
+```
+
+**String**
+
+```
+|------------|----------|---------|
+|  network   | protocol | payload |
+|------------|----------|---------|
+| 'f' or 't' |     0    | '0'-'9' |
```
Comment (whyrusleeping) [inline address.md]: probably put '0' in quotes to indicate we're putting it in as a string literal

Comment (whyrusleeping) [reply address.md]: also this is a little misleading. Reading this, I would expect the bytes for the payload to be the same for the raw and stringed versions, when its not. We should clarify that in the stringed version its base10 encoded

File: address.md (hunk 29 above)
Comment (whyrusleeping) [reply address.md]: We arent hashing anything, its just random bytes that don't mean anything.

Comment (whyrusleeping) [reply address.md]: (technically, we get those random bytes from hashing something, but thats an implementation detail that I think can be left out of the spec)

File: address.md (hunk 20 above)
Comment (whyrusleeping) [reply address.md]: New is a method on an address that returns an address?

File: address.md [hunk 35]
```diff
@@ -0,0 +1,300 @@
...
+Encodes an Address as a string, prepending the network prefix, calculating the checksum and encoding the payload and checksum to [base32](https://tools.ietf.org/html/rfc4648).
+
+```go
+func Encode(network string, a Address) string {
+	if network != "f" && network != "t" {
+        Fatal("Invalid Network")
+	}
+	
+    switch a.Protocol {
+    	case SECP256K1:
```
Comment (whyrusleeping) [inline address.md]: note that if using go-like syntax, this is broken. This function will no-op for Secp and Actor. either explicitly say 'fallthrough' or do `case BLS, Actor, SECP256k1:`

File: address.md [hunk 36]
```diff
@@ -0,0 +1,300 @@
...
+    if a[0] != "f" && a[0] != "t" {
+        Fatal(ErrUnknownNetwork)
+    }
+    
+    protocol := a[1]
+    raw := a[2:] 
+    if protocol == ID {
+        return Address{
+            Protocol: protocol,
+            Payload: raw,
```
Comment (whyrusleeping) [inline address.md]: this will need to be decoded from base10

File: address.md [hunk 37]
```diff
@@ -0,0 +1,300 @@
...
+    }
+    
+    payload = raw[:len(raw)-CksmLen]
+    if protocol == SECP256K1 || protocol == Actor {
+        if len(payload) != 20 {
+            Fatal(ErrInvalidBytes)
+        }
+    }
+    
+    cksm := payload[len(payload)-CksmLen : len(payload)]
```
Comment (whyrusleeping) [inline address.md]: protip:
`slice[6:]` means "from index 6 to the end"

File: address.md [hunk 38]
```diff
@@ -0,0 +1,300 @@
...
+    
+    protocol := a[1]
+    raw := a[2:] 
+    if protocol == ID {
+        return Address{
+            Protocol: protocol,
+            Payload: raw,
+        }
+    }
+
```
Comment (whyrusleeping) [inline address.md]: you will need to base32 decode these bits

File: address.md [hunk 39]
```diff
@@ -0,0 +1,300 @@
...
+	}
+	
+    switch a.Protocol {
+    	case SECP256K1:
+    	case Actor:
+    	case BLS:
+    		cksm := Checksum(a)
+    		return network + a.Protocol + base32.Encode(a.Payload + cksm)
+    	case ID:
+        	return network + a.Protocol + base32.Encode(a.Payload)
```
Comment (whyrusleeping) [inline address.md]: we don't use base32 encoding for addresses, we encode to base10

File: data-structures.md [hunk 40]
```diff
@@ -4,6 +4,24 @@ This document serves as an entry point for understanding all of the data structu
...
+    // 0: ID
+    // 1: SECP256K1 Public Key
+    // 2: Actor
+    // 3: BLS Public Key
+    protocol byte
+
+    // raw bytes containing the data associated with protocol
+    payload []byte
+}
+```
```
Comment (whyrusleeping) [inline data-structures.md]: link to the address spec here to discuss serialization.

File: address.md (hunk 30 above)
Comment (frrist) [reply address.md]: A user could use the same priv/pub key pair on the testnet as they do on the mainnet, this means the same actor could be referenced by the same address on either network (not advocating this, just a possibility). This is not true for ID addresses as they are not deterministically assigned.

This is probably more confusing than helpful, should I take it out?

File: address.md (hunk 33 above)
Comment (frrist) [reply address.md]: I was trying to indicate that the network and checksum won't always be present by have a "Filecoin address encoded to a byte" section and a "Filecoin address encoded to a string". What would make this more clear? Stating that only the bytes encoding will appear on chain?

File: address.md (hunk 20 above)
Comment (frrist) [reply address.md]: moved it off the interface

File: address.md [hunk 41]
```diff
@@ -0,0 +1,317 @@
...
+}
+```
+
+##### Encode()
+
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
```
Comment (whyrusleeping) [inline address.md]:
```suggestion
- produce an address with a valid checksum (if applicable)
```

File: address.md [hunk 42]
```diff
@@ -0,0 +1,317 @@
...
+	if network != "f" && network != "t" {
+		Fatal("Invalid Network")
+	}
+
+	switch a.Protocol {
+		case SECP256K1, Actor, BLS:
+			cksm := Checksum(a)
+			return network + a.Protocol + base32.Encode(a.Payload + cksm)
+		case ID:
+			return network + a.Protocol + base10.Encode(a.Payload)
```
Comment (whyrusleeping) [inline address.md]:
```suggestion
			return network + a.Protocol + base10.Encode(varint.Decode(a.Payload))
```

File: address.md [hunk 43]
```diff
@@ -0,0 +1,317 @@
...
+
+	protocol := a[1]
+	raw := a[2:]
+	if protocol == ID {
+		return Address{
+			Protocol: protocol,
+			Payload:  base10.Decode(raw),
+		}
+	}
+
```
Comment (whyrusleeping) [inline address.md]:
```suggestion
raw = base32.Decode(raw)
```

Comment (whyrusleeping) [reply address.md]: the checksum is base32 encoded as well, we have to decode the entire thing before slicing it off

File: address.md [hunk 44]
```diff
@@ -0,0 +1,317 @@
...
+	payload = raw[:len(raw)-CksmLen]
+	if protocol == SECP256K1 || protocol == Actor {
+		if len(payload) != 20 {
+			Fatal(ErrInvalidBytes)
+		}
+	}
+
+	cksm := base32.Decode(payload[len(payload)-CksmLen:])
```
Comment (whyrusleeping) [inline address.md]:
```suggestion
	cksm := payload[len(payload)-CksmLen:]
```

File: address.md [hunk 45]
```diff
@@ -0,0 +1,317 @@
...
+	if !ValidateChecksum(a, cksm) {
+		Fatal(ErrInvalidChecksum)
+	}
+
+	return Address{
+		Protocol: protocol,
+		Payload:  base32.Decode(payload),
```
Comment (whyrusleeping) [inline address.md]:
```suggestion
		Payload:  payload,
```

**Pull request merged**